rfcs/0100-sign-commits.md

```diff
@@ -0,0 +1,346 @@
+```bash
+#!/bin/sh
+# nixpkgsmerge
+
+git switch master
+git pull
+gh pr checkout $1
+git switch master
+git merge -m "Merged #$1" --no-ff --gpg-sign -
+git push origin master
+gh pr close $1
+```
+Then
+```bash
+nixpkgsmerge 123
+```
+
+For mobile platforms, you would use [nix-on-droid](https://github.com/t184256/nix-on-droid)
+to run the script instead.
+
+A merge commit must always be made so that there is something to sign.
+
+A git hook like the one presented here is also recommended,
+so that unsigned commits aren't pushed by accident.
+This would be put in `/.git/hooks/pre-push` of the repository to
+be committed to.
+```bash
+#!/bin/sh
+
+# Inspiration taken from https://github.com/guix-mirror/guix/blob/master/etc/git/pre-push.
+# Put this in .git/hooks/pre-push.
+set -xe
+
+# This is the "empty hash" used by Git when pushing a branch deletion.
+z40=0000000000000000000000000000000000000000
+
+while read local_ref local_hash remote_ref remote_hash
+do
+  # Branch deletion; no signature to verify.
+  if [ "$local_hash" != $z40 ]
+  then
+    # Check repositories under github.com/NixOS.
+    case "$2" in
+        *github.com*NixOS/*)
+      # TODO: More thorough check.
+      git verify-commit HEAD
+      ;;
+    esac
+  fi
+done
+```
+
+The digest algorithm for GPG (if you're using GPG) must be set to something that isn't SHA-1,
+as [it isn't secure anymore](https://eprint.iacr.org/2020/014.pdf), for example by putting the following in
+`$HOME/.gnupg/gpg.conf`:
+```
+digest-algo sha512
+```
+
+Committers will need to put their PGP keys in
+https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix,
+and add their keys to public key servers.
+
+The following might not be implemented before 2022-03-01T00:00Z,
+but the above will still regardless be required at that point in time.
+
+## Verifying flake Git inputs
+
+### Updating flake Git inputs
+[updating]: #updating-flake-git-inputs
+
+The design is inspired by what is described here: https://guix.gnu.org/en/blog/2020/securing-updates/.
+
+When updating any existing flake input of type `git` from an old commit, Nix must verify that
+the new commit is trusted.
+
+The old commit is always trusted.
+
+In the context of Git, a parent is a direct parent in the DAG of commits.
+
+If a commit has a trusted parent, then the commit is trusted if
+it doesn't have a trusted parent with `/.authorized-committers.nix`,
+
+If a commit has a trusted parent,
+and is signed by a committer listed in each of the trusted
+parents' `/.authorized-committers.nix`, then the commit is trusted.
+
+#### The format of `/.authorized-committers.nix`
+
+An example: `/.authorized-committers.nix`:
+```nix
+{
+  committers = {
+    jacob = "XXXX XXXX XXXX XXXX XXXX  XXXX XXXX XXXX XXXX XXXX";
+    jonathan = "YYYY YYYY YYYY YYYY YYYY  YYYY YYYY YYYY YYYY YYYY";
+  };
+}
+```
+
+Specifically, it will contain a Nix expression, that evaluates
```

Comment:
If it contains a nix expression it is already too late. At the point where you execute this code I'll have pirated a torrent via your land line as part of a fixed output derivation. We should be able to verify the integrity of the repository without executing code that is *within* the repository.

Reply:
You wouldn't be able to build the fixed-output derivation without recursive Nix or IFD AFAICT, but it might still be useful not to have to evaluate an untrusted Nix expression.

Reply:
It depends on how the eval code works in this case. How restrictive it is. It could probably be done in the same way a `flake.nix` is restricted from having thunks on the outer levels. This has to be explored and documented. If we can't do simple data transformations (that help with maintainability) of the signatures files then that is a major bummer.

Reply:
I already feel this major bummer with `flake.nix`, specifically you often duplicate the description text. This is something that will need a lot of thought IMO.

Reply:
Also this probably needs to be turned of by default as it is probably inedible slow on the entire tree.

Reply:
Doing a costly evaluation for each commit could indeed be annoying. Perhaps we should take the safe side for now and disallow computation.